### PR TITLE
Generalize access logging (rebased)

### PR DESCRIPTION
**Pull request by jrajahalme**
Refactor access logging internals from oxyproxy.go to accesslog/log.go, allowing for sharing most of the code with other proxy types.

Note that the (marked internal) http.Request has been removed from the access log message. This means that it is also removed from the json formatted output.

### CODE REVIEW

File: pkg/proxy/accesslog/log.go [hunk 1]
```diff
@@ -127,3 +204,102 @@ func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
...
+			info.Labels = c.Labels.Labels.GetModel()
+			info.LabelsSHA256 = c.Labels.GetLabelsSHA256()
+		}
+	}
+}
+
+// egressDestinationInfo returns the destination EndpointInfo for a flow
+// leaving the proxy at egress.
+func (info *EndpointInfo) egressDestinationInfo(ipstr string) {
+	ip := net.ParseIP(ipstr)
```
Comment (aanm) [inline pkg/proxy/accesslog/log.go]: I've seen this code in PR #1870 @manalibhutiyani or @jrajahalme is this duplicated or related?

Comment (tgraf) [reply pkg/proxy/accesslog/log.go]: We can rebase 1870 on top of this, 1870 is currently duplicating the http code for kafka

**Review by tgraf — APPROVED** [review]

File: pkg/proxy/accesslog/log.go [hunk 2]
```diff
@@ -88,33 +97,94 @@ func logString(outStr string, retry bool) {
...
+}
+
 // Log logs a record to the logfile and flushes the buffer
-func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
+func (l *LogRecord) Log(typ FlowType, verdict FlowVerdict, code int, info string) {
 	// Lock while writing access log so we serialize writes as we may have
 	// to reopen the logfile and parallel writes could fail because of that
+
+	l.Timestamp = time.Now().UTC().Format(time.RFC3339Nano)
+	l.Info = info
```
Comment (raybejjani) [inline pkg/proxy/accesslog/log.go]: Should these two writes be guarded by `logMutex`? This is the only place we use it and from the comment it seems to be a coordination lock more than a resource lock. If so, it might make sense to lock just before `logString` instead.

File: pkg/proxy/accesslog/log.go [hunk 3]
```diff
@@ -134,3 +204,102 @@ func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
...
+			info.LabelsSHA256 = c.Labels.GetLabelsSHA256()
+		}
+	}
+}
+
+// egressDestinationInfo returns the destination EndpointInfo for a flow
+// leaving the proxy at egress.
+func (info *EndpointInfo) egressDestinationInfo(ipstr string) {
+	ip := net.ParseIP(ipstr)
+	if ip != nil {
```
Comment (raybejjani) [inline pkg/proxy/accesslog/log.go]: This could be `if ip == nil { return }` if you wanted to reduce the indent by one.

**Review by raybejjani — CHANGES_REQUESTED** [review]
Another proxy owner should probably take a look too, but this is +1 overall. One thing about locks, however.

File: pkg/proxy/accesslog/log.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/proxy/accesslog/log.go]: These writes were done before taking the lock by the caller before this PR, so this PR just preserves the previous locking logic.

Looking closer I see the mutex is needed to serialize access to the log buffer and file. The caller owns the request, so there is no contention for accessing it. We could actually move the line taking a log a bit later in the Log() function.

File: pkg/proxy/accesslog/log.go [hunk 4]
```diff
@@ -17,10 +17,19 @@ package accesslog
...
 	"bufio"
 	"encoding/json"
+	"net"
+	"net/http"
+	"net/url"
 	"os"
+	"strconv"
+	"time"
 
+	"github.com/cilium/cilium/common/addressing"
```
Comment (rlenglet) [inline pkg/proxy/accesslog/log.go]: Package github.com/cilium/cilium/pkg/proxy/accesslog is a public API. It's used (vendored) by third-party software. Adding more transitive dependencies should be avoided. Please remove those additional dependencies. If that means moving the code depending on this out of github.com/cilium/cilium/pkg/proxy/accesslog, do that.

**Review by rlenglet — CHANGES_REQUESTED** [review]

File: pkg/proxy/accesslog/log.go (hunk 1 above)
Comment (manalibhutiyani) [reply pkg/proxy/accesslog/log.go]: I have removed the duplication actually in 1870

File: pkg/proxy/accesslog/log.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/proxy/accesslog/log.go]: Also noticed that OpenLogfile() and SetMetadata() did not take the lock. Not a problem currently, as they are only called from sync.Once(), but potential problem later, so I cleaned that up as well.

File: pkg/proxy/accesslog/log.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/proxy/accesslog/log.go]: OK, thanks!

File: pkg/proxy/accesslog/log.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/proxy/accesslog/log.go]: OK, was not aware of this. OK to remove the http.Request from the log record, though?

File: pkg/proxy/accesslog/log.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/proxy/accesslog/log.go]: I made the changes you requested, maybe review again?

**Review by rlenglet — APPROVED** [review]

**Review by tgraf — CHANGES_REQUESTED** [review]
#1824 is overlapping with this, please hold off on merging

File: pkg/proxy/accesslog/log.go (hunk 2 above)
Comment (raybejjani) [reply pkg/proxy/accesslog/log.go]: A+1 with extra credit :) Sorry for forcing you to track down all the callpaths to check locking :(

**Review by raybejjani — APPROVED** [review]

**Comment by jrajahalme** [comment]
Rebased to current master. Envoy integration depends on this, so please review :-)

**Review by rlenglet — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Review by manalibhutiyani — APPROVED** [review]
